Cannot un-ignore specific subdirectory in node_modules

I am developing React Native application and I am keeping the code inside ```node_modules/mylibrary```. The reason why I keep it there is to avoid relative paths between modules in my library and that's the recommended way to do that according to [this](https://github.com/substack/browserify-handbook#node_modules). I would use symlinks if it wasn't for [this](https://github.com/facebook/react-native/issues/637).

I would therefore like to lint code in ```node_modules/mylibrary```. Because ```node_modules``` are by default ignored, I've tried adding the following line to ```.eslintignore```, but it didn't help:
```
!node_modules/mylibrary/**/*.*
```

**What version of ESLint are you using?**
v2.2.0

**What configuration and parser (Espree, Babel-ESLint, etc.) are you using?**
My .eslintrc file looks like this:
```
{
  "extends": ["standard"],
  "parser": "babel-eslint"
}
```
**What did you do? Please include the actual source code causing the issue.**
I've created ```node_modules/mylibrary/abc.js``` with invalid content:
```
abc
```
Basically I am using atom editor and noticed that code is not linted. I confirmed it by calling three commands:
```
./node_modules/.bin/eslint --ignore-path .eslintignore ./node_modules/mylibrary
./node_modules/.bin/eslint ./node_modules/mylibrary
./node_modules/.bin/eslint --no-ignore ./node_modules/mylibrary
```
**What did you expect to happen?**
I would expect the output to be the same for all three commands

**What actually happened? Please include the actual, raw output from ESLint.**
First two commands output nothing (no errors). Third one output:
```
/Users/tomkur/git/myproject/node_modules/mylibrary/abc.js
  1:1  error  'abc' is not defined  no-undef

✖ 1 problem (1 error, 0 warnings)
```

--- a/lib/ignored-paths.js
+++ b/lib/ignored-paths.js
@@ -24,8 +24,8 @@ debug = debug("eslint:ignored-paths");
 
 var ESLINT_IGNORE_FILENAME = ".eslintignore";
 var DEFAULT_IGNORE_PATTERNS = [
-    "/node_modules/",
-    "/bower_components/"
+    "/node_modules/*",
+    "/bower_components/*"
 ];
 var DEFAULT_OPTIONS = {
     dotfiles: false,
@@ -106,6 +106,7 @@ function IgnoredPaths(options) {
     // Add a way to keep track of ignored files.  This was present in node-ignore
     // 2.x, but dropped for now as of 3.0.10.
     this.ig.custom.ignoreFiles = [];
+    this.ig.default.ignoreFiles = [];
 
     if (options.dotfiles !== true) {
         // ignore files beginning with a dot, but not files in a parent or ancestor
@@ -121,6 +122,7 @@ function IgnoredPaths(options) {
 
         if (options.ignorePattern) {
             addPattern(this.ig.custom, options.ignorePattern);
+            addPattern(this.ig.default, options.ignorePattern);
         }
 
         if (options.ignorePath) {
@@ -150,6 +152,7 @@ function IgnoredPaths(options) {
             debug("Adding " + ignorePath);
             this.baseDir = path.dirname(path.resolve(options.cwd, ignorePath));
             addIgnoreFile(this.ig.custom, ignorePath);
+            addIgnoreFile(this.ig.default, ignorePath);
         }
 
     }

--- a/lib/util/glob-util.js
+++ b/lib/util/glob-util.js
@@ -122,16 +122,20 @@ function listFilesToProcess(globPatterns, options) {
      */
     function addFile(filename, shouldWarnIgnored) {
         var ignored = false;
+        var isSilentlyIgnored;
         if (options.ignore !== false) {
             if (ignoredPaths.contains(filename, "default")) {
-                return;
+                isSilentlyIgnored = true;
             }
             if (ignoredPaths.contains(filename, "custom")) {
                 if (shouldWarnIgnored) {
                     ignored = true;
                 } else {
-                    return;
+                    isSilentlyIgnored = true;
                 }
+            }
+            if (isSilentlyIgnored && !ignored) {
+                return;
             }
         }
         if (added[filename]) {

--- a/tests/fixtures/ignored-paths/.eslintignoreWithUnignoredDefaults
+++ b/tests/fixtures/ignored-paths/.eslintignoreWithUnignoredDefaults
@@ -1,0 +1,2 @@
+!/node_modules/package
+!/bower_components/package

--- a/tests/lib/ignored-paths.js
+++ b/tests/lib/ignored-paths.js
@@ -14,7 +14,8 @@ var assert = require("chai").assert,
     os = require("os"),
     IgnoredPaths = require("../../lib/ignored-paths.js"),
     sinon = require("sinon"),
-    fs = require("fs");
+    fs = require("fs"),
+    includes = require("lodash").includes;
 
 require("shelljs/global");
 
@@ -36,7 +37,18 @@ function getIgnoreRules(ignoredPaths) {
     var ignoreRules = [];
 
     Object.keys(ignoredPaths.ig).forEach(function(key) {
-        ignoreRules = ignoreRules.concat(ignoredPaths.ig[key][ignoreRulesProperty]);
+        var rules = ignoredPaths.ig[key][ignoreRulesProperty];
+        rules.forEach(function(rule) {
+            var ruleOrigins = ignoreRules.map(function(ruleObj) {
+                return ruleObj.origin;
+            });
+            // Don't include duplicate ignore rules.
+            // (Duplicates occur because we add custom ignore patterns to the
+            // defaults as well, to allow unignoring default ignores)
+            if (!includes(ruleOrigins, rule.origin)) {
+                ignoreRules = ignoreRules.concat(rule);
+            }
+        });
     });
 
     return ignoreRules;
@@ -339,14 +351,14 @@ describe("IgnoredPaths", function() {
 
     describe("default ignores", function() {
 
-        it("should contain /bower_components/", function() {
+        it("should contain /bower_components/*", function() {
             var ignoredPaths = new IgnoredPaths();
-            assert.include(ignoredPaths.defaultPatterns, "/bower_components/");
+            assert.include(ignoredPaths.defaultPatterns, "/bower_components/*");
         });
 
-        it("should contain /node_modules/", function() {
+        it("should contain /node_modules/*", function() {
             var ignoredPaths = new IgnoredPaths();
-            assert.include(ignoredPaths.defaultPatterns, "/node_modules/");
+            assert.include(ignoredPaths.defaultPatterns, "/node_modules/*");
         });
 
         it("should always apply defaultPatterns if ignore option is true", function() {
@@ -367,6 +379,15 @@ describe("IgnoredPaths", function() {
             assert.isFalse(ignoredPaths.contains(getFixturePath("subdir/node_modules/package/file.js")));
         });
 
+        it("should allow subfolders of defaultPatterns to be unignored by ignorePattern", function() {
+            var ignoredPaths = new IgnoredPaths({ ignore: true, cwd: getFixturePath(), ignorePattern: "!/node_modules/package" });
+            assert.isFalse(ignoredPaths.contains(getFixturePath("node_modules", "package", "file.js")));
+        });
+
+        it("should allow subfolders of defaultPatterns to be unignored by ignorePath", function() {
+            var ignoredPaths = new IgnoredPaths({ ignore: true, cwd: getFixturePath(), ignorePath: getFixturePath(".eslintignoreWithUnignoredDefaults") });
+            assert.isFalse(ignoredPaths.contains(getFixturePath("node_modules", "package", "file.js")));
+        });
 
         it("should ignore dotfiles", function() {
             var ignoredPaths = new IgnoredPaths({ ignore: true, cwd: getFixturePath() });

--- a/tests/lib/util/glob-util.js
+++ b/tests/lib/util/glob-util.js
@@ -243,14 +243,14 @@ describe("globUtil", function() {
             assert.equal(result.length, 1);
         });
 
-        it("should not return a file listed in a specified ignore file", function() {
+        it("should not return a file from a glob if it matches a pattern in an ignore file", function() {
             var options = { ignore: true, ignorePath: getFixturePath("glob-util", "ignored", ".eslintignore") };
             var patterns = [getFixturePath("glob-util", "ignored", "**/*.js")];
             var result = globUtil.listFilesToProcess(patterns, options);
             assert.equal(result.length, 0);
         });
 
-        it("should not return a file listed in a specified ignore pattern", function() {
+        it("should not return a file from a glob if matching a specified ignore pattern", function() {
             var options = { ignore: true, ignorePattern: "foo.js", cwd: getFixturePath() };
             var patterns = [getFixturePath("glob-util", "ignored", "**/*.js")];
             var result = globUtil.listFilesToProcess(patterns, options);
@@ -272,6 +272,37 @@ describe("globUtil", function() {
             assert.deepEqual(result, [
                 {filename: file1, ignored: false}
             ]);
+        });
+
+        it("should set 'ignored: true' for files that are explicitly specified but ignored", function() {
+            var options = { ignore: true, ignorePattern: "foo.js", cwd: getFixturePath() };
+            var filename = getFixturePath("glob-util", "ignored", "foo.js");
+            var patterns = [filename];
+            var result = globUtil.listFilesToProcess(patterns, options);
+            assert.equal(result.length, 1);
+            assert.deepEqual(result, [
+                {filename: filename, ignored: true}
+            ]);
+        });
+
+        it("should not return files from default ignored folders", function() {
+            var options = { cwd: getFixturePath("glob-util") };
+            var glob = getFixturePath("glob-util", "**/*.js");
+            var patterns = [glob];
+            var result = globUtil.listFilesToProcess(patterns, options);
+            var resultFilenames = result.map(function(resultObj) {
+                return resultObj.filename;
+            });
+            assert.notInclude(resultFilenames, getFixturePath("glob-util", "node_modules", "dependency.js"));
+        });
+
+        it("should return unignored files from default ignored folders", function() {
+            var options = { ignorePattern: "!/node_modules/dependency.js", cwd: getFixturePath("glob-util") };
+            var glob = getFixturePath("glob-util", "**/*.js");
+            var patterns = [glob];
+            var result = globUtil.listFilesToProcess(patterns, options);
+            var unignoredFilename = getFixturePath("glob-util", "node_modules", "dependency.js");
+            assert.includeDeepMembers(result, [{filename: unignoredFilename, ignored: false}]);
         });
     });
 });

